smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/metrics/MetricDecorator.java

```diff
@@ -0,0 +1,39 @@
+package io.smallrye.reactive.messaging.metrics;
+
+import java.util.function.Consumer;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Instance;
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.metrics.Counter;
+import org.eclipse.microprofile.metrics.MetricRegistry;
+import org.eclipse.microprofile.metrics.Tag;
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.eclipse.microprofile.reactive.streams.operators.PublisherBuilder;
+
+import io.smallrye.reactive.messaging.PublisherDecorator;
+
+@ApplicationScoped
+public class MetricDecorator implements PublisherDecorator {
+
+    @Inject
+    private Instance<MetricRegistry> registryInstance;
+
+    @Override
+    public PublisherBuilder<? extends Message> decoratePublisher(PublisherBuilder<? extends Message> publisher,
+            String channelName) {
+        if (registryInstance.isResolvable()) {
+            return publisher.peek(incrementCount(channelName));
+        } else {
+            return publisher;
+        }
+    }
+
+    private <T extends Message> Consumer<T> incrementCount(String channelName) {
+        MetricRegistry registry = registryInstance.get();
+        Counter counter = registry.counter("mp.messaging.message.count", new Tag("channel", channelName));
```

Comment:
Should we be creating a new `Tag` instance every time, or should we be caching them?

Could we even cache the `Counter` instance by channel name? `registry.counter()` internally calls a `synchronized` method to retrieve the counter, so we probably want to minimize calls to it?

Reply:
This code only gets called once for each channel, the last line returns the lambda that gets called for each message.

Reply:
So `publisher.peek()` is a continual operation?

The javadoc reads like it only operates on the current state of the publisher, which to me implies any items that are added to the stream after `peek()` is called wouldn't have this applied.

I'm still grokking reactive so I could be way off base with my thinking

Reply:
Yes, the consumer passed to `peek()` is called for each element emitted from the publisher. It's like a `map()` which doesn't alter the input.